### ISSUE TITLE
Add any BATCH file in autoexec config section breaks autolaunch DOSEXE

### ISSUE THREAD

**Issue by DSkywalk** [dosbox-staging/dosbox-staging#4018]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.82.0

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Ubuntu 20.04

### Relevant hardware info

Standard Linux PC

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

The combination of `@echo off` and any BATCH file in the `[autoexec]` freezes dosbox if you run a game from the command line.

![imagen](https://github.com/user-attachments/assets/90055c48-4d0b-40b7-9770-81b033f3066d)

Documentation says:
```
                           If PATH is a DOS executable (BAT/COM/EXE), it's parent
                           path is mounted as C: and the executable is run. When
                           the executable exits, DOSBox Staging quits.
```

But this doesn't  work if the user try to run a BATCH file, on init... As a workaround the user can use `CALL`.

_This problem appear even running an empty BATCH file._

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Create an empty TEST.BAT and add to `[autoexec]` section with an `@echo off`.
2. DOSBox starts normally by default, but...
3. Try to exec a program from the CLI then DOSBox keep in a black screen with a cursor blinking.

### Download URL of affected game or software

[dosbox-bug-hang.zip](https://github.com/user-attachments/files/17544435/DOSBOX-BUG-REPORT.zip)

### Your configuration

```ini
[autoexec]
# Each line in this section is executed at startup as a DOS command.
@echo off
TEST.BAT
```
and run: ` $ dosbox GAME.EXE`

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by interloper98** [comment]
@DSkywalk , I checked where that documentation is coming from, and it's regarding arguments to the dosbox staging executable itself.


**`dosbox --help`**

```
Usage: dosbox [OPTION]... [PATH]                                    

PATH   If PATH is a directory, it's mounted as C:.
       If PATH is a bootable disk image (IMA/IMG), it's booted.
       If PATH is a CD-ROM image (CUE/ISO), it's mounted as D:.
       If PATH is a DOS executable (BAT/COM/EXE), it's parent
       path is mounted as C: and the executable is run. When
       the executable exits, DOSBox Staging quits.
```

So in your case, if you want to follow the documentation, delete all the stuff you added to [autoexec].

Outside of dosbox, create a `TEST.BAT` file (add line to it, like: `@echo hello world`).

Let's say you saved this in your Linux directory here: `/home/dskywalk/dos/test/TEST.BAT`

Now launch dosbox with the path to your new BAT file. 

`dosbox.exe ~/dos/test/TEST.BAT`

According to the docs, DOSBox Staging will mount `~/dos/test` as C: and then proceed to launch your batch file. 

Hopefully it works 🙂

**Comment by DSkywalk** [comment]
Sorry, Maybe I haven't explained myself well.

The problem is not running a BAT from autoexec, the problem is that if you run a BAT file in autoexec and you want to run dosbox with a game, so that it launches it directly when you run, it hangs.

```
[autoexec]
# Each line in this section is executed at startup as a DOS command.
@echo off
TEST.BAT
```
And now you run a game from cmd:

```
dosbox /foo/path/to/game/START.EXE
```

dosbox-staging crashes, for me is a bug :+1:

**Comment by DSkywalk** [comment]
I tested other forks like dosbox-x and when launch a game from directly ignores your `[autoexec]` section, I doesn't like this solution but at least not hangs it :face_exhaling:

**Comment by johnnovak** [comment]
@DSkywalk Ok, so the behaviour is a bit confusing, and there is at least one bug here.

DOSBox doesn't "hang" in this case, just when you have something like this in your local `dosbox.conf`

```ini
[autoexec]
@echo off
test.bat
```

and you execute `dosbox blah.exe` from that folder where the config resides, echoing stays off. This is probably a bug because if you just start `dosbox` (without passing in the name of an executable as the argument), echoing is on.

You can confirm this in your "hanging" test case: just type the name of the executable blind, and press enter. The executable will start, so the problem is the echoing "stuck" in disabled state.

I have to say, this is a pretty unorthodox way to use DOSBox and an edge case, but technically probably a bug 😆

**Comment by johnnovak** [comment]
> Sorry again to copy my `[autoexec]` section, I had at the beginning the `@echo off` but this is indifferent to the problem, it happens exactly the same.

Exact repro steps please. It's hard to follow what exactly are you referring to now.

Please just ZIP up a simple example in its entirety and also include the launch commands you're using.

**Comment by DSkywalk** [comment]
A really good idea, also doing some more tests I have found another error, if you call a BAT in the autoexec the execution doesn't work either.

[DOSBOX-BUG-REPORT.zip](https://github.com/user-attachments/files/17544435/DOSBOX-BUG-REPORT.zip)

**Comment by interloper98** [comment]
Poked around with this and found there are a couple things going on.

`echo` normally prints its arguments to the screen, however in a batch file `echo off` and `echo on` are actually special commands that control the DOS shell's printing of parsed commands. `echo off` will turn off command printing indefinitely for the life of the batch file until you turn it back on again (with `echo on`). 

If you want to only hide an individual command in batch file, you can prefix it with `@`.  `@WOLF.EXE` will launch wolf but not show you the command on the command line.

If you're 100% confident in your batch file and want it to run without any echoing.. then add `@echo off` to the top of the BAT, and it will run silent and clean (all dark!)

I searched the conf and found a couple things that are in play here, that I think explain it.

```
# startup_verbosity: Controls verbosity prior to displaying the program ('auto' by default):
#    Verbosity   | Welcome | Early stdout
#    high        |   yes   |    yes
#    low         |   no    |    yes
#    quiet       |   no    |    no
#
#    auto   'low' if exec or dir is passed, otherwise 'high'
#           Possible values: auto, high, low, quiet.
```

It says when a executable or directory is passed (presumably on the dosbox command line), then we get the low verbosity. I tried this:

dosbox --set startup_verbosity=high -conf bugged-autoexec.conf GAME/WOLF3D.EXE

And yes, I now see it mounting `T:\` and running TEST.BAT where there's nothing left to do.

---

But what about WOLF3D - why didn't it run it after we called TEST.BAT? 

Let's test this in real MS-DOS. I've created three .BAT files (and left echo default on, so we're going to see every command on the shell).

![read-dos-test](https://github.com/user-attachments/assets/11ce61da-a0f3-48ca-aeb5-72e82ac9e710)

First I run `nocall.bat` that **_runs_** `test.bat`. 

![real-dos-nocall-test](https://github.com/user-attachments/assets/c0d690fa-e6f2-4026-85d1-0a993139b3d5)

Notice how we never get back out of `test.bat` because the `"back out at nocall"` message is never printed. We're stuck inside (even `exit` doesn't get back to the parent batch file!).

Now let's try `usecall.bat` that **_calls_** `test.bat`. This time we actually get back out!

![real-dos-usecall-test](https://github.com/user-attachments/assets/6b2876d2-f5d6-46ed-b27f-2d25a0c21d15)

So when I look at what DOSBox Staging is doing (getting "stuck" inside TEST.BAT because we **_ran_** instead of **_call_**ed it -- yes, it's doing the same thing exactly like MS-DOS does).

Switching it to `call test.bat` then launches WOLF3D correctly.

The key point here is that CALL will resume execution in the parent shell at the point of the CALL, once the CALLED batch finishes. Where as not using call doesn't resume. Running diverts execution instead of nesting it (the parent is effectively lost).

---

> I think the only logical solution is to always execute the autoexec section, and if you provide an exe name in the command line, then execute that after the commands from the autoexec.

Also confirmed that, @johnnovak.

Looks like there's this setting that does that be default:

```
#  autoexec_section: How autoexec sections are handled from multiple config files:
#    join:       Combine them into one big section (legacy behavior; default).
#    overwrite:  Use the last one encountered, like other config settings.
#                Possible values: join, overwrite.
```

I tested this, and when I use `overwrite`, I can 100% clobber prior conf's `[autoexec]` content with later confs.. and finally passing in an executable is the final clobber. 

So if I do this:

dosbox --set autoexec_section=overwrite -conf bugged-autoexec.conf GAME/WOLF3D.EXE


It just launches WOLF3D because I guess that's the final autoexec content, and clobbers the [autoexec] from the conf.

**Comment by DSkywalk** [comment]
Also the behaviour with `echo off` from normal prompt: 

![imagen](https://github.com/user-attachments/assets/01d900d3-4b5f-441e-8897-717d4589f9ac)

Using: `./dosbox-bug-hang.sh`
![imagen](https://github.com/user-attachments/assets/cf423c81-6bc1-4617-a8a9-50e85c7c8e88)

**Comment by interloper98** [comment]
Yeah, looks like real MS-DOS resets the `@echo off` state at the end of batch file, where as in Staging it looks like it's still active:

![echo-off](https://github.com/user-attachments/assets/54954b76-10ac-4aac-bb61-7e1fdc08c44c)

When I run the same locally in Staging, it matches this:

![noecho-staging](https://github.com/user-attachments/assets/f4616eea-524a-4e48-8de7-53fae077b86e)

However, when it's run through the `[autoexec]` system it seems like the `@echo off` state isn't release (maybe we're "in" the top-most shell and there isn't anything to exit out too, unlike when running them from the CLI?).

I guess I can try doing the same in MS-DOS's real autoexec.bat :-)

**Comment by interloper98** [comment]
@DSkywalk,

Launch with `--set startup_verbosity=low`:

`dosbox --set startup_verbosity=low -conf bugged-autoexec.conf GAME/WOLF3D.EXE

It then matches real MS-DOS.

It looks like that documentation isn't right:

```
# startup_verbosity: Controls verbosity prior to displaying the program ('auto' by default):
#    Verbosity   | Welcome | Early stdout
#    high        |   yes   |    yes
#    low         |   no    |    yes
#    quiet       |   no    |    no
#
#    auto   'low' if exec or dir is passed, otherwise 'high'
#           Possible values: auto, high, low, quiet.
```

When an executable is passed in, it's actually is using `quiet` instead of `low`.

So I think that's a bug. We should use `low` instead of `quiet` and match the docs.

Probably a quick fix with some grepping about :laughing: will submit it later when I have a minute.

**Comment by DSkywalk** [comment]
Yes, I have seen that this way you recover control again, but it doesn't launch the executable which it is supposed to do as well... maybe we should separate both processes and not just add at the end of the `[autoexec]` the commands to run the game :thinking:

**Comment by arrowgent** [comment]
hmmm its an interesting approach and discovery specifically

i will just note, a lot of "how to" suggest the simple task of "drag & drop the exe onto dosbox" solution.  which is effectively the same as running the command `dosbox game.exe` from a terminal/cmd.
ie: click and drag the game exe onto dosbox exe

this holds true in linux and windows im sure, although linux is more prone to going directly to a terminal to execute things rather than dragging and dropping


i can see examples such as GOG which may have an [autoexec] .bat that is supposed to run the game.  if someone were to replace the provided dosbox of the gog game with dosbox-staging then tried the same thing; it may not run correctly.

just some thoughts.

**Comment by johnnovak** [comment]
@DSkywalk Can you please confirm this has been fixed in the latest dev build?

**Comment by BigRedTower** [comment]
Hi, I believe this is normal and proper behavior. I'm using Windows but I'm pretty sure the same applies under Linux.

In DOS .BAT files and modern Windows .CMD files, when launching a nested batch file you need to use **CALL** with the child batch filename.

I was able to get both .conf files in DOSBOX-BUG-REPORT.zip to work as expected by adding CALL.

In the `[autoexec]` section of the .conf files, instead of:
`TEST.BAT`

Use:
`CALL TEST.BAT`

And you will get the expected behavior: Wolf3D starts normally in both dosbox-bug-hang and dosbox-bug-wrong-init.

This appears to be normal behavior, to me. Without the CALL command, .bat/.cmd interpreters abandon the parent batch script and run the child batch script, and never return control to the parent batch script. With CALL, the child batch script is run, then execution resumes in the parent batch script after the child script completes.

When launching a .BAT file from the [autoexec] section, always use CALL, since [autoexec] defines an AUTOEXEC.BAT file.

Hope this helps!

**Comment by johnnovak** [comment]
Fixed by https://github.com/dosbox-staging/dosbox-staging/pull/4054
